test/validator.go

```diff
@@ -1,197 +1,273 @@
 package test
 
 import (
+	"context"
+	"encoding/json"
 	"fmt"
-	"os"
-	"path/filepath"
+	"strconv"
+	"testing"
+	"time"
 
-	cometcrypto "github.com/cometbft/cometbft/crypto"
-	cometcryptoed25519 "github.com/cometbft/cometbft/crypto/ed25519"
+	cometbytes "github.com/cometbft/cometbft/libs/bytes"
+	cometjson "github.com/cometbft/cometbft/libs/json"
 	"github.com/cometbft/cometbft/privval"
-	"github.com/ory/dockertest"
-	"github.com/strangelove-ventures/horcrux/signer"
+	"github.com/cosmos/cosmos-sdk/types/bech32"
+	slashingtypes "github.com/cosmos/cosmos-sdk/x/slashing/types"
+	"github.com/docker/docker/client"
+	"github.com/strangelove-ventures/horcrux/signer/proto"
+	interchaintest "github.com/strangelove-ventures/interchaintest/v7"
+	"github.com/strangelove-ventures/interchaintest/v7/chain/cosmos"
+	"github.com/strangelove-ventures/interchaintest/v7/ibc"
+	"github.com/strangelove-ventures/interchaintest/v7/testutil"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
 )
 
-type Validator struct {
-	Index     int
-	Sentries  map[string]Nodes
-	Signers   Signers
-	tl        Logger
-	Home      string
-	PubKeys   map[string]cometcrypto.PubKey
-	Threshold uint8
-}
+const (
+	testChain        = "gaia" // ghcr.io/strangelove-ventures/heighliner/gaia
+	testChainVersion = "v10.0.2"
 
-func NewHorcruxValidator(
-	tl Logger,
-	pool *dockertest.Pool,
-	networkID string,
-	home string,
-	index int,
-	numSigners int,
-	threshold uint8,
-	chains ...*ChainType,
-) (*Validator, error) {
-	chainIDs := make([]string, 0, len(chains))
-	sentries := make(map[string]Nodes)
-	for _, chain := range chains {
-		sentries[chain.ChainID] = MakeNodes(
-			index,
-			chain.NumSentries,
-			home,
-			chain.ChainID,
-			chain,
-			pool,
-			networkID,
-			tl,
-		)
-		chainIDs = append(chainIDs, chain.ChainID)
-	}
-	testValidator := &Validator{
-		Index:     index,
-		Sentries:  sentries,
-		Signers:   MakeSigners(index, numSigners, home, pool, networkID, tl),
-		tl:        tl,
-		Home:      home,
-		Threshold: threshold,
-		PubKeys:   make(map[string]cometcrypto.PubKey),
-	}
-	if err := testValidator.genPrivKeyAndShards(nil, chainIDs...); err != nil {
-		return nil, err
-	}
-	return testValidator, nil
+	signerPort       = "2222"
+	signerPortDocker = signerPort + "/tcp"
+
+	signerImage        = "horcrux-test"
+	binary             = "horcrux"
+	signerImageUidGid  = "2345:2345"
+	signerImageHomeDir = "/home/horcrux"
+)
+
+// chainWrapper holds the initial configuration for a chain to start from genesis.
+type chainWrapper struct {
+	chain           **cosmos.CosmosChain
```

Comment:
The double star (pointer to a pointer) is uncommon. Have you tried using the `chainWrapper` strust as a pointer when callers use it? This may eliminate the double pointer.

Reply:
The double star removed the need for double wrapped functions for pre-genesis, but I updated it so that it's the single pointer.